mac/Keyman4MacIM/Keyman4MacIM/KMInputController.m

```diff
@@ -32,95 +67,391 @@ - (id)initWithServer:(IMKServer *)server delegate:(id)delegate client:(id)inputC
 }
 
 - (NSUInteger)recognizedEvents:(id)sender {
-    return (NSKeyDownMask | NSLeftMouseDownMask);
+    return (NSKeyDownMask | NSLeftMouseDownMask | NSLeftMouseUpMask);
 }
 
 - (BOOL)handleEvent:(NSEvent *)event client:(id)sender {
     if ([self.AppDelegate debugMode])
         NSLog(@"Event = %@", event);
     
-    if (event == nil || sender == nil)
-        return NO;
+    if (event == nil || sender == nil || self.kmx == nil)
+        return NO; // Not sure this can ever happen.
     
     // OSK key feedback from hardware keyboard is disabled
     /*if (event.type == NSKeyDown)
         [self.AppDelegate handleKeyEvent:event];*/
     
-    if (self.kmx == nil)
+    if (event.type == NSLeftMouseDown || event.type == NSLeftMouseUp ) {
+//        if (_clientSelectionCanChangeUnexpectedly) {
+//            if ([self.AppDelegate debugMode])
+//                NSLog(@"WARNING: We are dealing with an app/context where we THINK we shouldn't be getting mouse events, but we just got one!");
+//            _clientSelectionCanChangeUnexpectedly = NO;
+//        }
+        _contextOutOfDate = YES;
         return NO;
+    }
+    else if (event.type != NSKeyDown)
+        return NO; // We ignore NSLeftMouseDragged events (because we'll eventually get a mouse-up).
+    else if ((event.modifierFlags & NSEventModifierFlagCommand) == NSEventModifierFlagCommand) {
+        // There are a bunch of common navigation/selection command-key combinations, but individual
+        // apps may implement specifc commands that also chnage the selection. There is probably no
+        // situation where a user could reasonably hope that any dead-keys typed before using a
+        // command shortcut would be remembered, so other than an insignificant performance penalty,
+        // the only downside to treating all commands as having the potential to change the selection
+        // is that some "legacy" apps can't get their context at all.
+        if ([self.AppDelegate debugMode])
+            NSLog(@"Command key - context needs to be re-gotten.");
+        _contextOutOfDate = YES;
+        return NO; // We let the client app handle all Command-key events.
+    }
     
-    if (event.type == NSLeftMouseDown) {
-        [self performSelector:@selector(updateContextBuffer:) withObject:sender afterDelay:0.25];
-        return NO;
+    if (_legacyMode && event.keyCode == kProcessPendingBuffer)
+    {
+        if ([self.AppDelegate debugMode])
```

Comment:
Can we refactor this into something like `[self.AppDelegate] debugLog(message, parameters);`? Not sure what the correct notation is? Or is it difficult with format strings?

Reply:
Maybe. I was just following the lead set by the existing code. It might make for slightly cleaner code, but for the common (non-debug) case it would add a bunch of unnecessary function calls that would just be no-ops. Maybe that's something to consider in a follow-up refactoring pass. I'd like to try to do some refactoring anyway.